build/templates/session.py.mako

```diff
@@ -273,7 +283,15 @@ class Session(_SessionBase):
 
         ${helper.get_function_docstring(init_function, False, config, indent=8)}
         '''
-        super(Session, self).__init__(repeated_capability_list=[], ${config['session_handle_parameter_name']}=None, library=None, encoding=None, freeze_it=False)
+        # Initialize the superclass with default values first, populate them later
+        super(Session, self).__init__(
+            repeated_capability_list=[],
```

Comment:
If we're doing this, then why make them parameters of `__init__()` at all?
Alternatively, and probably cleaner, can we invoke super's `__init__()` later?

Reply:
This is actually the approach used by the existing code, so I'm not too sure about the actual rationale behind.

Based on my understanding, these have to be parameters of `_SessionBase.__init__()` because it may be called from here (i.e. `Session.__init__()`) and `_RepeatedCapabilities.__getitem__()` (which will actually call it with the final values instead of default values).

We couldn't invoke super's `__init__()` later because that requires calling `self.get_channel_names()` and `self.channel_count` first to get the final `all_channels_in_session`, but calling these method and attribute requires repeated capabilities related properties to be populated first, which is currently done in `_SessionBase.__init__()`, so it kinda creates a circular dependency. If we really want to invoke super's `__init__()` later, we'll need to duplicate the logic in `_SessionBase.__init__()` to `Session.__init__()`, which is not desirable in my opinion.

Hence, I suggest that we retain the current approach (my added comment should have made it clear enough that the initial values passed to the super class are just default values that would be overridden).

Reply:
> This is actually the approach used by the existing code

The fact that the existing code does things in some way does not mean it is the most correct way or that it doesn't have room for improvement.

If anything this further highlights how tightly coupled the implementations of `_SessionBase` and `Session` are, which is not great architecturally but also not something we can fix cheaply.